beacon_node/network/src/subnet_service/mod.rs

```diff
@@ -17,12 +32,645 @@ pub enum SubnetServiceMessage {
+    // interpretation of this variable seemed like the best way to group the logic, even though it
+    // may be counter-intuitive (apologies to future readers).
+    pub slot: Slot,
+}
+
+/// The enum used to group all kinds of validator subscriptions
+#[derive(Debug, Clone, PartialEq)]
+pub enum Subscription {
+    Attestation(ValidatorSubscription),
+    SyncCommittee(SyncCommitteeSubscription),
+}
+
+pub struct SubnetService<T: BeaconChainTypes> {
+    /// Queued events to return to the driving service.
+    events: VecDeque<SubnetServiceMessage>,
+
+    /// A reference to the beacon chain to process received attestations.
+    pub(crate) beacon_chain: Arc<BeaconChain<T>>,
+
+    /// Subnets we are currently subscribed to as short lived subscriptions.
+    ///
+    /// Once they expire, we unsubscribe from these.
+    /// We subscribe to subnets when we are an aggregator for an exact subnet.
+    // NOTE: When setup the default timeout is set for sync committee subscriptions.
+    subscriptions: HashSetDelay<Subnet>,
+
+    /// Subscriptions that need to be executed in the future.
+    scheduled_subscriptions: HashSetDelay<Subnet>,
+
+    /// A list of permanent subnets that this node is subscribed to.
+    // TODO: Shift this to a dynamic bitfield
+    permanent_attestation_subscriptions: HashSet<Subnet>,
+
+    /// A collection timeouts to track the existence of aggregate validator subscriptions at an
+    /// `ExactSubnet`.
+    aggregate_validators_on_subnet: Option<HashSetDelay<ExactSubnet>>,
+
+    /// The waker for the current thread.
+    waker: Option<std::task::Waker>,
+
+    /// The discovery mechanism of lighthouse is disabled.
+    discovery_disabled: bool,
+
+    /// We are always subscribed to all subnets.
+    subscribe_all_subnets: bool,
+
+    /// Whether this node is a block proposer-only node.
+    proposer_only: bool,
+
+    /// The logger for the attestation service.
+    log: slog::Logger,
+}
+
+impl<T: BeaconChainTypes> SubnetService<T> {
+    /* Public functions */
+
+    /// Establish the service based on the passed configuration.
+    pub fn new(
+        beacon_chain: Arc<BeaconChain<T>>,
+        node_id: NodeId,
+        config: &NetworkConfig,
+        log: &slog::Logger,
+    ) -> Self {
+        let log = log.new(o!("service" => "attestation_service"));
+
+        let slot_duration = beacon_chain.slot_clock.slot_duration();
+
+        if config.subscribe_all_subnets {
+            slog::info!(log, "Subscribing to all subnets");
+        }
+
+        // Build the list of known permanent subscriptions, so that we know not to subscribe or
+        // discover them.
+        let mut permanent_attestation_subscriptions = HashSet::default();
+        if config.subscribe_all_subnets {
+            // We are subscribed to all subnets, set all the bits to true.
+            for index in 0..<T::EthSpec as EthSpec>::SubnetBitfieldLength::to_u64() {
```

Comment:
We should probably be using `ATTESTATION_SUBNET_COUNT` from config instead even though this is unlikely to change.

Reply:
yep